Services/Mail/classes/class.ilMailExplorer.php

```diff
@@ -1,111 +1,371 @@
+     * @param $parentObject
+     * @param $userId
      */
-    public function __construct($a_parent_obj, $a_parent_cmd, $a_user_id)
+    public function __construct($parentObject, $userId)
     {
         global $DIC;
 
-        $this->lng         = $DIC->language();
-        $this->ctrl        = $DIC->ctrl();
-        $this->httpRequest = $DIC->http()->request();
+        $this->lng          = $DIC->language();
+        $this->ctrl         = $DIC->ctrl();
+        $this->httpRequest  = $DIC->http()->request();
+        $this->parentObject = $parentObject;
+        $this->ui           = $DIC->ui();
 
-        $this->tree = new ilTree($a_user_id);
+        $this->tree = new ilTree($userId);
         $this->tree->setTableNames('mail_tree', 'mail_obj_data');
+    }
 
-        parent::__construct('mail_exp', $a_parent_obj, $a_parent_cmd, $this->tree);
+    /**
+     * Get Tree UI
+     *
+     * @return Tree|object
+     */
+    public function getTreeComponent()
+    {
+        $f = $this->ui->factory();
+        /** @var ilTree $tree */
+        $tree = $this->tree;
+
+        $subtree  = $tree->getChilds($tree->readRootId());
+        $data     = $subtree;
 
-        $this->setSkipRootNode(true);
-        $this->setAjax(false);
-        $this->setOrderField('title,m_type');
+        $tree = $f->tree()
+                  ->expandable($this)
+                  ->withData($data)
+                  ->withHighlightOnNodeClick(true);
+
+        return $tree;
     }
 
-    function getNodeContent($a_node)
+    /**
+     * Get a list of records (that list can also be empty).
+     * Each record will be relayed to $this->build to retrieve a Node.
+     * Also, each record will be asked for Sub-Nodes using this function.
+     * @param      $record
+     * @param null $environment
+     * @return array
+     */
+    public function getChildren($record, $environment = null) : array
     {
-        $content = $a_node['title'];
+        return $this->getChildsOfNode($record['child']);
+    }
 
-        if ($a_node['child'] == $this->getNodeId($this->getRootNode())) {
-            $content = $this->lng->txt('mail_folders');
-        } elseif ($a_node['depth'] < 3) {
-            $content = $this->lng->txt('mail_' . $a_node['title']);
+    /**
+     * Build and return a Node.
+     * The renderer will provide the $factory-parameter which is the UI-factory
+     * for nodes, as well as the (unspecified) $environment as configured at the Tree.
+     * $record is the data the node should be build for.
+     *
+     * @param Factory $factory
+     * @param         $record
+     * @param null    $environment
+     * @return Node
+     */
+    public function build(
+        Factory $factory,
+        $record,
+        $environment = null
+    ) : Node {
+        $node = $this->createNode($factory, $record);
+
+        $href = $this->getNodeHref($record);
+
+        if ($href) {
+            $node = $node->withAdditionalOnLoadCode( function($id) use ($href) {
```

Comment:
Do we really need this? We should use a `$node->withLink($uri);` instead, right?